barrier.go

```diff
@@ -0,0 +1,273 @@
+}
+func (n *idleBarrier) BatchPoint(m edge.BatchPointMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		n.resetTimer()
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *idleBarrier) EndBatch(m edge.EndBatchMessage) (edge.Message, error) {
+	return m, nil
+}
+func (n *idleBarrier) Barrier(m edge.BarrierMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		n.resetTimer()
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *idleBarrier) DeleteGroup(m edge.DeleteGroupMessage) (edge.Message, error) {
+	if m.GroupID() == n.group.ID {
+		n.Stop()
+	}
+	return m, nil
+}
+
+func (n *idleBarrier) Point(m edge.PointMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		n.resetTimer()
+		return m, nil
+	}
+	return nil, nil
+}
+
+func (n *idleBarrier) resetTimer() {
+	if n.idle > 0 {
+		n.timer.Reset(n.idle)
+	}
+}
+
+func (n *idleBarrier) emitBarrier() error {
+	nowT := time.Now()
+	n.lastT.Store(nowT)
+	return edge.Forward(n.outs, edge.NewBarrierMessage(n.group, nowT))
+}
+
+func (n *idleBarrier) idleHandler() {
+	defer n.wg.Done()
+	for {
+		select {
+		case <-n.timer.C:
+			n.emitBarrier()
+			n.resetTimer()
+		case <-n.stopC:
+			return
+		}
+	}
+}
+
+type periodicBarrier struct {
+	name  string
+	group edge.GroupInfo
+
+	lastT  atomic.Value
+	ticker *time.Ticker
+	wg     sync.WaitGroup
+	outs   []edge.StatsEdge
+	stopC  chan bool
+}
+
+func newPeriodicBarrier(name string, group edge.GroupInfo, period time.Duration, outs []edge.StatsEdge) *periodicBarrier {
+	r := &periodicBarrier{
+		name:   name,
+		group:  group,
+		lastT:  atomic.Value{},
+		ticker: time.NewTicker(period),
+		wg:     sync.WaitGroup{},
+		outs:   outs,
+		stopC:  make(chan bool, 1),
+	}
+
+	r.Init()
+
+	return r
+}
+
+func (n *periodicBarrier) Init() {
+	n.lastT.Store(time.Time{})
+	n.wg.Add(1)
+
+	go n.periodicEmitter()
+}
+
+func (n *periodicBarrier) Stop() {
+	n.stopC <- true
+	n.ticker.Stop()
+	n.wg.Wait()
+}
+
+func (n *periodicBarrier) BeginBatch(m edge.BeginBatchMessage) (edge.Message, error) {
+	return m, nil
+}
+func (n *periodicBarrier) BatchPoint(m edge.BatchPointMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *periodicBarrier) EndBatch(m edge.EndBatchMessage) (edge.Message, error) {
+	return m, nil
+}
+func (n *periodicBarrier) Barrier(m edge.BarrierMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		return m, nil
+	}
+	return nil, nil
+}
+func (n *periodicBarrier) DeleteGroup(m edge.DeleteGroupMessage) (edge.Message, error) {
+	if m.GroupID() == n.group.ID {
+		n.Stop()
+	}
+	return m, nil
+}
+
+func (n *periodicBarrier) Point(m edge.PointMessage) (edge.Message, error) {
+	if !m.Time().Before(n.lastT.Load().(time.Time)) {
+		return m, nil
+	}
+	return nil, nil
+}
+
+func (n *periodicBarrier) emitBarrier() error {
+	nowT := time.Now()
```

Comment:
This time needs to be the last time seen in the data stream, not the current time. This is because its possible to write historical data, i.e. in a replay.

Reply:
Replay already seems to work.  My only reservation with using last message time here is that we'd need to store the last message time for every received message, which, if we're using a Mutex to protect access to the variable, would mean we're acquiring a write lock for every time a message is received.  If we can continue using atomic rather than Mutex this might be OK.  Also I have to admit I don't know what the cost of Mutex is in go vs atomic operations.

Reply:
Using atomic to store the last time is fine, you can even use an int64 as nanoseconds since the epoch. You should only need to store the greatest time per group.

In general atomic is faster than getting a mutex lock, but atomic only works if each individual piece of state can be modified independently, otherwise you need a mutex lock to ensure all state is modified together. But in either case making an atomic write or getting a mutex lock shouldn't be much cost compared to the cost of processing messages in general.

Reply:
So beyond the performance impact (which I agree is negligible) it seems like setting barrier time to the last time seen in the data stream creates a problem when trying to handle a stream with no data.

My thought was that window should be forced to emit when a barrier is received with a timestamp that is past the nextEmit time, which would make sense from the window's perspective since no data past the time in the barrier will be received, therefore it can assume that the window needs to be emitted.  However, setting barrier time to the last time seen in the data stream makes it such that the window would either
1) never emit until a message past nextEmit is received, which makes it so that silent periods can't be handled, or
2) emit any time a barrier is received, irregardless of whether the barrier time is before nextEmit, which logically doesn't make sense to me

So the first would bring us back to square one since we need the ability to force emission of window when there are silent periods.  So I think to support the forced emit behavior either barrier time gets set to the wall clock or we force window to emit on receipt of barrier irregardless of timestamp on the barrier...  thoughts?